File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> temp;
        vector<vector<int>> ans;
        sort(nums.begin(),nums.end());
        set<vector<int>> st;
        
        helper(0,temp,st,nums);
        
        for(auto i : st){
            ans.push_back(i);
        }
        
        return ans;
    }
    
    void helper(int idx,vector<int>& temp,set<vector<int>>& st,vector<int>& nums){
        if(idx == nums.size()){
            st.insert(temp);
            return;
        }
        
        temp.push_back(nums[idx]);
        helper(idx+1,temp,st,nums);
        temp.pop_back();
        helper(idx+1,temp,st,nums);
    }
};

File: 0090-subsets-ii/0090-subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []

        def dfs(i,arr):
            if i == len(nums):
                ans.append(arr.copy())
                return
            
            arr.append(nums[i])
            dfs(i+1,arr)
            arr.pop()

            while i+1 < len(nums) and nums[i+1] == nums[i]:
                i += 1
            
            dfs(i+1,arr)
        
        dfs(0,[])

        return ans